### ISSUE TITLE
Allow variable declarations with guard-for-in

### ISSUE THREAD

**Issue by jvarho** [eslint/eslint#7398]
**What rule do you want to change?**
[guard-for-in](http://eslint.org/docs/rules/guard-for-in)

**Does this change cause the rule to produce more or fewer warnings?**
Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior.

**Please provide some example code that this change will affect:**

``` js
for (key in foo) {
    const val = checkSomething(key)
    if (val > 0) {
        doSomething(key, val);
    }
}
```

**What does the rule currently do for this code?**
Warns.

**What will the rule do after it's changed?**
Not warn.

---

The point is that the condition(s) that you check to filter properties may involve values that are also needed inside the loop body. It would be good to allow this without having to disable or work around the warning. I.e. allow `let` and `const` statements before the `if`.

**Comment by not-an-aardvark** [comment]
Thanks for the suggestion.

I'm not sure about this change; I think the point of the `guard-for-in` rule is to enforce the following pattern:

```js
for (key in foo) {
  // Only do things if `key` is an own property of `foo`, rather than an inherited property.
  if (Object.prototype.hasOwnProperty.call(foo, key)) {
    // do logic, including const and let declarations, in this `if` statement
  }
}
```

**Comment by jvarho** [comment]
There are other valid filters and you may want to combine it with something. For example:

```javascript
for (key in foo) {
  const bar = foo[key]
  if (Object.prototype.hasOwnProperty.call(foo, key) && Array.isArray(bar)) {
    // do logic using bar
  }
}
```

**Comment by platinumazure** [comment]
@jvarho Well, in that case, the `foo[key]` access is occurring without first having validated that `key` is an own property. That is exactly what the rule is designed to warn about.

You have a couple of options:
1. Write the `for` loop as suggested by the rule.
1. Disable the rule for that particular expression.

Writing the loop per the rule:

```js
for (key in foo) {
    if (Object.prototype.hasOwnProperty.call(foo, key)) {
        const bar = foo[key];
        if (Array.isArray(bar)) {
            // etc.
        }
    }
}
```

Disabling the rule:

```js
// eslint-disable-next-line guard-for-in
for (key in foo) {
    const bar = foo[key];
    if (Array.isArray(bar)) {
        // etc.
    }
}
```

**Comment by jvarho** [comment]
Well, I disagree that the rule should warn about such clearly correct code, but I don't feel strongly about it so feel free to close if this will not be changed.

**Comment by kaicataldo** [comment]
Closing this as discussion has stalled and it doesn't appear there is enough interest, either from the team or from the community, to continue the discussion at this time. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. Please feel free to open a new issue with a proposal to continue this discussion if this is something you would like to continue advocating for.

**Comment by jpickwell** [comment]
I would like to not get warnings about not having a for-in guard when I have a `let` statement before the guard. Example of what should be correct code:
```javascript
for(const key in obj)
{
  let value

  if(!Object.prototype.hasOwnProperty.call(obj, key))
  {
    continue
  }

  value = obj[key]

  // ... conditional code to potentially re-assign value
}
```

Notice that property access is still guarded.

**Comment by kaicataldo** [comment]
Feel free to make a new issue with a proposal!
